File: src/main/java/com/minecraft/moonlake/enums/Enchantment.java
/*
 * Copyright (C) 2016 The MoonLake Authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */


package com.minecraft.moonlake.enums;

import com.minecraft.moonlake.MoonLakeAPI;
import com.minecraft.moonlake.api.utility.MinecraftVersion;
import com.minecraft.moonlake.exception.IllegalBukkitVersionException;

import javax.annotation.Nullable;

/**
 * <h1>Enchantment</h1>
 * 附魔类型中文汉化版（详细doc待补充...）
 *
 * @version 1.1
 * @author Month_Light
 */
public abstract class Enchantment {

    // 翻译均来自中文 wiki
    // http://minecraft-zh.gamepedia.com/%E9%99%84%E9%AD%94#.E9.AD.94.E5.92.92
    ///

    /**
     * 附魔类型: 保护
     */
    public final static Enchantment 保护 = new EnchantmentWrapped(0, "PROTECTION_ENVIRONMENTAL");
    /**
     * 附魔类型: 火焰保护
     */
    public final static Enchantment 火焰保护 = new EnchantmentWrapped(1, "PROTECTION_FIRE");
    /**
     * 附魔类型: 摔落保护
     */
    public final static Enchantment 摔落保护 = new EnchantmentWrapped(2, "PROTECTION_FALL");
    /**
     * 附魔类型: 爆炸保护
     */
    public final static Enchantment 爆炸保护 = new EnchantmentWrapped(3, "PROTECTION_EXPLOSIONS");
    /**
     * 附魔类型: 弹射物保护
     */
    public final static Enchantment 弹射物保护 = new EnchantmentWrapped(4, "PROTECTION_PROJECTILE");
    /**
     * 附魔类型: 水下呼吸
     */
    public final static Enchantment 水下呼吸 = new EnchantmentWrapped(5, "OXYGEN");
    /**
     * 附魔类型: 水下速掘
     */
    public final static Enchantment 水下速掘 = new EnchantmentWrapped(6, "WATER_WORKER");
    /**
     * 附魔类型: 荆棘
     */
    public final static Enchantment 荆棘 = new EnchantmentWrapped(7, "THORNS");
    /**
     * 附魔类型: 深海探索者
     */
    public final static Enchantment 深海探索者 = new EnchantmentWrapped(8, "DEPTH_STRIDER");

    /**
     * 附魔类型: 锋利
     */
    public final static Enchantment 锋利 = new EnchantmentWrapped(16, "DAMAGE_ALL");
    /**
     * 附魔类型: 亡灵杀手
     */
    public final static Enchantment 亡灵杀手 = new EnchantmentWrapped(17, "DAMAGE_UNDEAD");
    /**
     * 附魔类型: 节肢杀手
     */
    public final static Enchantment 节肢杀手 = new EnchantmentWrapped(18, "DAMAGE_ARTHROPODS");
    /**
     * 附魔类型: 击退
     */
    public final static Enchantment 击退 = new EnchantmentWrapped(19, "KNOCKBACK");
    /**
     * 附魔类型: 火焰附加
     */
    public final static Enchantment 火焰附加 = new EnchantmentWrapped(20, "FIRE_ASPECT");
    /**
     * 附魔类型: 抢夺
     */
    public final static Enchantment 抢夺 = new EnchantmentWrapped(21, "LOOT_BONUS_MOBS");

    /**
     * 附魔类型: 效率
     */
    public final static Enchantment 效率 = new EnchantmentWrapped(32, "DIG_SPEED");
    /**
     * 附魔类型: 精准采集
     */
    public final static Enchantment 精准采集 = new EnchantmentWrapped(33, "SILK_TOUCH");
    /**
     * 附魔类型: 耐久
     */
    public final static Enchantment 耐久 = new EnchantmentWrapped(34, "DURABILITY");
    /**
     * 附魔类型: 时运
     */
    public final static Enchantment 时运 = new EnchantmentWrapped(35, "LOOT_BONUS_BLOCKS");

    /**
     * 附魔类型: 力量
     */
    public final static Enchantment 力量 = new EnchantmentWrapped(48, "ARROW_DAMAGE");
    /**
     * 附魔类型: 冲击
     */
    public final static Enchantment 冲击 = new EnchantmentWrapped(49, "ARROW_KNOCKBACK");
    /**
     * 附魔类型: 火矢
     */
    public final static Enchantment 火矢 = new EnchantmentWrapped(50, "ARROW_FIRE");
    /**
     * 附魔类型: 无限
     */
    public final static Enchantment 无限 = new EnchantmentWrapped(51, "ARROW_INFINITE");

    /**
     * 附魔类型: 海之眷顾
     */
    public final static Enchantment 海之眷顾 = new EnchantmentWrapped(61, "LUCK");
    /**
     * 附魔类型: 饵钓
     */
    public final static Enchantment 饵钓 = new EnchantmentWrapped(62, "LURE");

    // 1.9 增加的新附魔（不兼容 1.8 以下
    /**
     * 附魔类型: 冰霜行者 (1.9+)
     */
    public final static Enchantment 冰霜行者 = new EnchantmentWrapped(9, "FROST_WALKER", MinecraftVersion.V1_9);
    /**
     * 附魔类型: 经验修补 (1.9+)
     */
    public final static Enchantment 经验修补 = new EnchantmentWrapped(70, "MENDING", MinecraftVersion.V1_9);
    ///

    // 1.11 增加的新附魔（不兼容 1.10 以下
    /**
     * 附魔类型: 绑定诅咒 (1.11+)
     */
    public final static Enchantment 绑定诅咒 = new EnchantmentWrapped(10, "BINDING_CURSE", MinecraftVersion.V1_11);
    /**
     * 附魔类型: 消失诅咒 (1.11+)
     */
    public final static Enchantment 消失诅咒 = new EnchantmentWrapped(71, "VANISHING_CURSE", MinecraftVersion.V1_11);
    ///

    // 1.11.1 增加的新附魔（不兼容 1.11 以下
    /**
     * 附魔类型: 横扫之刃 (1.11.1+)
     */
    public final static Enchantment 横扫之刃 = new EnchantmentWrapped(22, "SWEEPING_EDGE", new MinecraftVersion(1, 11, 1));
    ///

    private final int id;
    private final String name;
    private final MinecraftVersion requiredVersion;

    /**
     * 附魔类型类构造函数
     *
     * @param id 附魔 ID
     * @param name 附魔名称
     */
    private Enchantment(int id, String name) {

        this(id, name, null);
    }

    /**
     * 附魔类型类构造函数
     *
     * @param id 附魔 ID
     * @param name 附魔名称
     * @param requiredVersion 需求版本
     */
    private Enchantment(int id, String name, MinecraftVersion requiredVersion) {

        this.id = id;
        this.name = name;
        this.requiredVersion = requiredVersion;
    }

    /**
     * 获取此附魔类型的 Id
     *
     * @return Id
     * @deprecated 已过时, 请使用 {@link #getName()}
     */
    @Deprecated
    public int getId() {

        return id;
    }

    /**
     * 获取此附魔类型的名称
     *
     * @return 名称
     */
    public String getName() {

        return name;
    }

    /**
     * 获取此附魔类型需求的次版本号
     *
     * @return 次版本号
     * @deprecated 已过时, 将于 v2.0 删除. 请使用 {@link #getRequiredMCVersion()}
     */
    @Deprecated
    public int getRequiredVersion() { // TODO 2.0

        return requiredVersion != null ? requiredVersion.getMinor() : -1;
    }

    /**
     * 获取此附魔类型需求的版本
     *
     * @return 版本
     */
    @Nullable
    public MinecraftVersion getRequiredMCVersion() {

        return requiredVersion;
    }

    /**
     * 转换为 Bukkit 的 Enchantment 对象
     *
     * @return Enchantment
     * @throws IllegalBukkitVersionException 如果 Bukkit 服务器版本不支持附魔则抛出异常
     */
    public org.bukkit.enchantments.Enchantment as() {

        if(requiredVersion != null && !MoonLakeAPI.currentMCVersion().isLater(requiredVersion)) {

            throw new IllegalBukkitVersionException("The bukkit version not support enchantment: " + id);
        }
        //return org.bukkit.enchantments.Enchantment.getById(id);
        return org.bukkit.enchantments.Enchantment.getByName(name);
    }

    private final static class EnchantmentWrapped extends Enchantment {

        private EnchantmentWrapped(int id, String name) {
            super(id, name);
        }

        private EnchantmentWrapped(int id, String name, MinecraftVersion requiredVersion) {
            super(id, name, requiredVersion);
        }
    }
}

File: API/src/main/kotlin/com/minecraft/moonlake/api/item/Enchantment.kt
/*
 * Copyright (C) 2016-2017 The MoonLake (mcmoonlake@hotmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.minecraft.moonlake.api.item

import com.minecraft.moonlake.api.Valuable
import com.minecraft.moonlake.api.adapter.AdapterBukkit
import com.minecraft.moonlake.api.version.MinecraftVersion

enum class Enchantment(val id: Int, val max: Int, val type: String, val mcVer: MinecraftVersion? = null) : AdapterBukkit<org.bukkit.enchantments.Enchantment>, Valuable<String> {

    PROTECTION(0, 4, "PROTECTION_"),                                                                  	   保护(PROTECTION),
    PROTECTION_FIRE(1, 4, "PROTECTION_FIRE"),                                                        火焰保护(PROTECTION_FIRE),
    PROTECTION_FALL(2, 4, "PROTECTION_FALL"),                                                      摔落保护(PROTECTION_FALL),
    PROTECTION_EXPLOSIONS(3, 4, "PROTECTION_EXPLOSIONS"),                            爆炸保护(PROTECTION_EXPLOSIONS),
    PROTECTION_PROJECTILE(4, 4, "PROTECTION_PROJECTILE"),                                 弹射物保护(PROTECTION_PROJECTILE),
    OXYGEN(5, 3, "OXYGEN"),                                                                         			   水下呼吸(OXYGEN),
    WATER_WORKER(6, 1, "WATER_WORKER"),                                                             水下速倔(WATER_WORKER),
    THORNS(7, 3, "THORNS"),                                                                         			   荆棘(THORNS),
    DEPTH_STRIDER(8, 3, "DEPTH_STRIDER"),                                                                深海探索者(DEPTH_STRIDER),
    FROST_WALKER(9, 2, "FROST_WALKER", MinecraftVersion.V1_9),                            冰霜行者(FROST_WALKER),
    BINDING_CURSE(10, 1, "BINDING_CURSE", MinecraftVersion.V1_11),                    绑定诅咒(BINDING_CURSE),

    DAMAGE(16, 5, "DAMAGE_ALL"),                                                                      	   锋利(DAMAGE),
    DAMAGE_UNDEAD(17, 5, "DAMAGE_UNDEAD"),                                                   亡灵杀手(DAMAGE_UNDEAD),
    DAMAGE_ARTHROPODS(18, 5, "DAMAGE_ARTHROPODS"),                                   节肢杀手(DAMAGE_ARTHROPODS),
    KNOCKBACK(19, 2, "KNOCKBACK"),                                                                	   击退(KNOCKBACK),
    FIRE_ASPECT(20, 2, "FIRE_ASPECT"),                                                                 	   火焰附加(FIRE_ASPECT),
    LOOT_BONUS_MOBS(21, 3, "LOOT_BONUS_MOBS"),                                            抢夺(LOOT_BONUS_MOBS),
    SWEEPING_EDGE(22, 3, "SWEEPING_EDGE", MinecraftVersion(1, 11, 1)),                横扫之刃(SWEEPING_EDGE),

    DIG_SPEED(32, 5, "DIG_SPEED"),                                                                   			效率(DIG_SPEED),
    SILK_TOUCH(33, 1, "SILK_TOUCH"),                                                                 		精准采集(SILK_TOUCH),
    DURABILITY(34, 3, "DURABILITY"),                                                                  			耐久(DURABILITY),
    LOOT_BONUS_BLOCKS(35, 3, "LOOT_BONUS_BLOCKS"),                                       时运(LOOT_BONUS_BLOCKS),

    ARROW_DAMAGE(48, 5, "ARROW_DAMAGE"),                                                        力量(ARROW_DAMAGE),
    ARROW_KNOCKBACK(49, 2, "ARROW_KNOCKBACK"),                                             冲击(ARROW_KNOCKBACK),
    ARROW_FIRE(50, 1, "ARROW_FIRE"),                                                                 		火矢(ARROW_FIRE),
    ARROW_INFINITE(51, 1, "ARROW_INFINITE"),                                                          无限(ARROW_INFINITE),

    LUCK(61, 3, "LUCK"),                                                                             					海至眷顾(LUCK),
    LURE(62, 3, "LURE"),                                                                              					饵钓(LURE),

    MENDING(70, 1, "MENDING", MinecraftVersion.V1_9),                                 			经验修补(MENDING),
    VANISHING_CURSE(71, 1, "VANISHING_CURSE", MinecraftVersion.V1_11),           	消失诅咒(VANISHING_CURSE),
    ;

    constructor(equivalent: Enchantment) : this(equivalent.id, equivalent.max, equivalent.type, equivalent.mcVer)

    override fun value(): String
            = type

    /**
     * @throws IllegalArgumentException if the level is unsafe
     */
    @Throws(IllegalArgumentException::class)
    fun checkSafe(level: Int) // TODO check item enchantment support
            = if(level < 1 || level > max) throw IllegalArgumentException() else Unit

    override fun cast(): org.bukkit.enchantments.Enchantment
            = org.bukkit.enchantments.Enchantment.getByName(type)

    companion object {

        @JvmStatic
        private val ID_MAP: MutableMap<Int, Enchantment> = HashMap()

        init {
            val regex = Regex("(i?)[a-z]([a-z_\\d]*)")
            values().forEach { if(it.name.matches(regex)) ID_MAP.put(it.id, it) }
        }

        @JvmStatic
        @JvmName("fromName")
        fun fromName(name: String): Enchantment
                = Enchantment.valueOf(name.toUpperCase())

        @JvmStatic
        @JvmName("fromId")
        @Throws(IllegalArgumentException::class)
        fun fromId(id: Int): Enchantment
                = ID_MAP[id] ?: throw IllegalArgumentException("错误: 无效的附魔效果 ID $id 值.")

        @JvmStatic
        @JvmName("hasId")
        fun hasId(id: Int): Boolean
                = ID_MAP.containsKey(id)
    }
}
